Adding algorithm TD Learning with N-Tuple Networks for 2048

Jazeem opened the pull request:
An algorithm that can win at 2048 without using tree search or hand crafted state evaluator functions.

Currently the code is somewhat coupled with 2048. It is possible to decouple it, if that is valuable for open_spiel. Will have to find a way to move n_tuples into game.

Have modified the game 2048 to consider a move that doesn't change the board as invalid

Implementation is similar to https://github.com/moporgic/TDL2048-Demo

lanctot commented:
Cool! We had quite a big refactor committed to the master branch yesterday. Could you pull changes from master and commit the merge (or rebase)?

We basically just moved all the games into subdirectories.

I cannot run the tests while there are conflicts.

Jazeem commented:
Sure. Rebased now 👍🏻

lanctot commented on open_spiel/python/examples/2048_td_n_tuple_network.py:
```diff
@@ -0,0 +1,120 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""TD Learning with N-Tuple Networks for 2048"""
```
Can you put a brief description here too of what this example does / shows / outputs?

lanctot commented on open_spiel/python/examples/2048_td_n_tuple_network.py:
```diff
@@ -0,0 +1,120 @@
+    "Episode frequency at which the agent is evaluated.")
+FLAGS = flags.FLAGS
+
+n_tuple_size = 6
+max_tuple_index = 15
+n_tuples = [[0, 1, 2, 3, 4, 5],[4, 5, 6, 7, 8, 9],
+    [0, 1, 2, 4, 5, 6],[4, 5, 6, 8, 9, 10],]
+n_tuple_network_size = len(n_tuples)
+
+look_up_table_shape = (
+    n_tuple_network_size,) + (max_tuple_index,) * n_tuple_size
+look_up_table = np.zeros(look_up_table_shape)
+alpha = 0.02
```
Make these global or put them in main. (or make them flags)

lanctot commented on open_spiel/python/examples/2048_td_n_tuple_network.py:
```diff
@@ -0,0 +1,120 @@
+    n_tuple_network_size,) + (max_tuple_index,) * n_tuple_size
+look_up_table = np.zeros(look_up_table_shape)
+alpha = 0.02
+
```
Two lines between globals and function (e.g. main)

lanctot commented on open_spiel/python/examples/2048_td_n_tuple_network.py:
```diff
@@ -0,0 +1,120 @@
+            sum_rewards = 0
+            largest_tile = 0
+            max_score = 0
+
```
here too, two blank lines. Similarly below

lanctot reviewed (changes requested):
Can you lint your code by following Step 9 in the Adding a New Game on this page? https://github.com/deepmind/open_spiel/blob/master/docs/developer_guide.md

It greatly reduced the load on our end to have the code formatted as required by Google python standards. 

E.g. indentation should be two spaces, not four. Etc.

Jazeem commented:
@lanctot 

* Used pylint and formatted the code
* Introduced a class NTupleNetwork and added a brief description about what N-Tuples are and a citation to a IEEE paper that this implementation is based on
* The output of the code is very straightforward similar to breakthrough_dqn.py in that for every 1000 runs; the code prints what's the average score reached, max score and largest tile that was unlocked. Running the code you can see the average scores increasing indicating that learning is happening. Once the 2048 tile is unlocked, the agent is able to win the game

lanctot commented:
Hi @Jazeem,

Can you respond to the conversations above (or resolve them now if they are resolved)

Jazeem commented:
@lanctot 

Have resolved them
